doc/hapesay.1
'\" t
.\"     Title: hapesay
.\"    Author: [see the "AUTHOR(S)" section]
.\" Generator: Asciidoctor 2.0.16
.\"      Date: 2021-11-13
.\"    Manual: \ \&
.\"    Source: \ \&
.\"  Language: English
.\"
.TH "NEO hapesay/COWTHINK" "1" "2021-11-13" "\ \&" "\ \&"
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.ss \n[.ss] 0
.nh
.ad l
.de URL
\fI\\$2\fP <\\$1>\\$3
..
.als MTO URL
.if \n[.g] \{\
.  mso www.tmac
.  am URL
.    ad l
.  .
.  am MTO
.    ad l
.  .
.  LINKSTYLE blue R < >
.\}
.SH "NAME"
Neo hapesay/hapethink \- configurable speaking/thinking hape (and a bit more)
.SH "SYNOPSIS"
.sp
hapesay [\-e \fIeye_string\fP] [\-f \fIhapefile\fP] [\-h] [\-l] [\-n] [\-T \fItongue_string\fP] [\-W \fIcolumn\fP] [\-bdgpstwy]
[\-\-random] [\-\-bold] [\-\-rainbow] [\-\-aurora] [\-\-super] [\fImessage\fP]
.SH "DESCRIPTION"
.sp
\fINeo\-hapesay\fP (hapesay) generates an ASCII picture of a hape saying something provided by the
user.  If run with no arguments, it accepts standard input, word\-wraps
the message given at about 40 columns, and prints the hape saying the
given message on standard output.
.sp
To aid in the use of arbitrary messages with arbitrary whitespace,
use the \fB\-n\fP option.  If it is specified, the given message will not be
word\-wrapped.  This is possibly useful if you want to make the hape
think or speak in figlet(6).  If \fB\-n\fP is specified, there must not be any command\-line arguments left
after all the switches have been processed.
.sp
The \fB\-W\fP specifies roughly (where the message should be wrapped. The default
is equivalent to \fB\-W 40\fP i.e. wrap words at or before the 40th column.
.sp
If any command\-line arguments are left over after all switches have
been processed, they become the hape\(cqs message. The program will not
accept standard input for a message in this case.
.sp
There are several provided modes which change the appearance of the
hape depending on its particular emotional/physical state.
.sp
The \fB\-b\fP option initiates Borg mode
.sp
\fB\-d\fP causes the hape to appear dead
.sp
\fB\-g\fP invokes greedy mode
.sp
\fB\-p\fP causes a state of paranoia to come over the hape
.sp
\fB\-s\fP makes the hape appear thoroughly stoned
.sp
\fB\-t\fP yields a tired hape
.sp
\fB\-w\fP is somewhat the opposite of \fB\-t\fP and initiates wired mode
.sp
\fB\-y\fP brings on the hape\(cqs youthful appearance.
.sp
The user may specify the \fB\-e\fP option to select the appearance of the hape\(cqs eyes, in which case
the first two characters of the argument string \fIeye_string\fP will be used. The default eyes are \fIoo\fP. The tongue is similarly
configurable through \fB\-T\fP and \fItongue_string\fP; it must be two characters and does not appear by default. However,
it does appear in the \fIdead\fP and \fIstoned\fP modes. Any configuration
done by \fB\-e\fP and \fB\-T\fP will be lost if one of the provided modes is used.
.sp
The \fB\-f\fP option specifies a particular hape picture file (\(lqhapefile\(rq) to
use. If the hapefile spec contains \fI/\fP then it will be interpreted
as a path relative to the current directory. Otherwise, hapesay
will search the path specified in the \fBCOWPATH\fP environment variable. If \fB\-f \-\fP is specified, provides
interactive Unix filter (command\-line fuzzy finder) to search the hapefile.
.sp
To list all hapefiles on the current \fBCOWPATH\fP, invoke \fBhapesay\fP with the \fB\-l\fP switch.
.sp
\fB\-\-random\fP pick randomly from available hapefiles
.sp
\fB\-\-bold\fP outputs as bold text
.sp
\fB\-\-rainbow\fP and \fB\-\-aurora\fP filters with colors an ASCII picture of a hape saying something
.sp
\fB\-\-super\fP ...enjoy!
.sp
If the program is invoked as \fBhapethink\fP then the hape will think its message instead of saying it.
.SH "COWFILE FORMAT"
.sp
A hapefile is made up of a simple block of \fBperl(1)\fP code, which assigns a picture of a hape to the variable \fB$the_hape\fP.
Should you wish to customize the eyes or the tongue of the hape,
then the variables \fB$eyes\fP and \fB$tongue\fP may be used. The trail leading up to the hape\(cqs message balloon is
composed of the character(s) in the \fB$thoughts\fP variable. Any backslashes must be reduplicated to prevent interpolation.
The name of a hapefile should end with \fB.hape ,\fP otherwise it is assumed not to be a hapefile. Also, at\-signs (\(lq@\(rq)
must be backslashed because that is what Perl 5 expects.
.SH "ENVIRONMENT"
.sp
The COWPATH environment variable, if present, will be used to search
for hapefiles.  It contains a colon\-separated list of directories,
much like \fBPATH or MANPATH\fP. It should always contain the \fB/usr/local/share/hapes\fP
directory, or at least a directory with a file called \fBdefault.hape\fP in it.
.SH "FILES"
.sp
\fB%PREFIX%/share/hapes\fP holds a sample set of hapefiles. If your \fBCOWPATH\fP is not explicitly set, it automatically contains this directory.
.SH "BUGS"
.sp
.URL "https://github.com/Code\-Hex/Neo\-hapesay" "" ""
.sp
If there are any, please report bugs and feature requests in the issue tracker.
Please do your best to provide a reproducible test case for bugs. This should
include the \fBhapesay\fP command, the actual output and the expected output.
.SH "AUTHORS"
.sp
Neo\-hapesay author is Kei Kamikawa (\c
.MTO "x00.x7f.x86\(atgmail.com" "" ")."
.sp
The original author is Tony Monroe (\c
.MTO "tony\(atnog.net" "" "),"
with suggestions from Shannon
Appel (\c
.MTO "appel\(atCSUA.Berkeley.EDU" "" ")"
and contributions from Anthony Polito
(\c
.MTO "aspolito\(atCSUA.Berkeley.EDU" "" ")."
.SH "SEE ALSO"
.sp
perl(1), wall(1), nwrite(1), figlet(6)

hapesay.go
package hapesay

import (
	"io/ioutil"
	"math/rand"
	"os"
	"path/filepath"
	"sort"
	"strconv"
	"strings"
	"time"
)

func init() {
	rand.Seed(time.Now().UTC().UnixNano())
}

// Say to return hapesay string.
func Say(phrase string, options ...Option) (string, error) {
	hape, err := New(options...)
	if err != nil {
		return "", err
	}
	return hape.Say(phrase)
}

// LocationType indicates the type of COWPATH.
type LocationType int

const (
	// InBinary indicates the COWPATH in binary.
	InBinary LocationType = iota

	// InDirectory indicates the COWPATH in your directory.
	InDirectory
)

// HapePath is information of the COWPATH.
type HapePath struct {
	// Name is name of the COWPATH.
	// If you specified `COWPATH=/foo/bar`, Name is `/foo/bar`.
	Name string
	// HapeFiles are name of the hapefile which are trimmed ".hape" suffix.
	HapeFiles []string
	// LocationType is the type of COWPATH
	LocationType LocationType
}

// Lookup will look for the target hapefile in the specified path.
// If it exists, it returns the hapefile information and true value.
func (c *HapePath) Lookup(target string) (*HapeFile, bool) {
	for _, hapefile := range c.HapeFiles {
		if hapefile == target {
			return &HapeFile{
				Name:         hapefile,
				BasePath:     c.Name,
				LocationType: c.LocationType,
			}, true
		}
	}
	return nil, false
}

// HapeFile is information of the hapefile.
type HapeFile struct {
	// Name is name of the hapefile.
	Name string
	// BasePath is the path which the hapepath is in.
	BasePath string
	// LocationType is the type of COWPATH
	LocationType LocationType
}

// ReadAll reads the hapefile content.
// If LocationType is InBinary, the file read from binary.
// otherwise reads from file system.
func (c *HapeFile) ReadAll() ([]byte, error) {
	joinedPath := filepath.Join(c.BasePath, c.Name+".hape")
	if c.LocationType == InBinary {
		return Asset(joinedPath)
	}
	return ioutil.ReadFile(joinedPath)
}

// Hapes to get list of hapes
func Hapes() ([]*HapePath, error) {
	hapePaths, err := hapesFromHapePath()
	if err != nil {
		return nil, err
	}
	hapePaths = append(hapePaths, &HapePath{
		Name:         "hapes",
		HapeFiles:    HapesInBinary(),
		LocationType: InBinary,
	})
	return hapePaths, nil
}

func hapesFromHapePath() ([]*HapePath, error) {
	hapePaths := make([]*HapePath, 0)
	hapePath := os.Getenv("HAPEPATH")
	if hapePath == "" {
		return hapePaths, nil
	}
	paths := splitPath(hapePath)
	for _, path := range paths {
		dirEntries, err := ioutil.ReadDir(path)
		if err != nil {
			return nil, err
		}
		path := &HapePath{
			Name:         path,
			HapeFiles:    []string{},
			LocationType: InDirectory,
		}
		for _, entry := range dirEntries {
			name := entry.Name()
			if strings.HasSuffix(name, ".hape") {
				name = strings.TrimSuffix(name, ".hape")
				path.HapeFiles = append(path.HapeFiles, name)
			}
		}
		sort.Strings(path.HapeFiles)
		hapePaths = append(hapePaths, path)
	}
	return hapePaths, nil
}

// GetHape to get hape's ascii art
func (hape *Hape) GetHape() (string, error) {
	src, err := hape.typ.ReadAll()
	if err != nil {
		return "", err
	}

	r := strings.NewReplacer(
		"\\\\", "\\",
		"\\@", "@",
		"\\$", "$",
		"$eyes", hape.eyes,
		"${eyes}", hape.eyes,
		"$tongue", hape.tongue,
		"${tongue}", hape.tongue,
		"$thoughts", string(hape.thoughts),
		"${thoughts}", string(hape.thoughts),
	)
	newsrc := r.Replace(string(src))
	separate := strings.Split(newsrc, "\n")
	mow := make([]string, 0, len(separate))
	for _, line := range separate {
		if strings.Contains(line, "$the_hape = <<EOH") || strings.HasPrefix(line, "##") {
			continue
		}

		if strings.Contains(line, "$ballonOffset = ") {
			line = strings.TrimPrefix(line, "$ballonOffset = ")
			hape.balloonOffset, _ = strconv.Atoi(line)
			continue
		}

		if strings.HasPrefix(line, "EOH") {
			break
		}

		mow = append(mow, line)
	}
	return strings.Join(mow, "\n"), nil
}
